crates/sparrow-execution/src/lib.rs

```diff
@@ -8,254 +8,193 @@
-
-    #[tokio::test]
-    #[ignore]
-    async fn test_query() {
-        sparrow_testing::init_test_logging();
-
-        let (input_tx, input_rx) = tokio::sync::mpsc::channel(10);
-        let (output_tx, mut output_rx) = tokio::sync::mpsc::channel(10);
-
-        let input_fields = Fields::from(vec![
-            Field::new("a", DataType::Int64, true),
-            Field::new("b", DataType::Int64, true),
-            Field::new("c", DataType::Int64, true),
-        ]);
-
-        let output_fields = Fields::from(vec![
-            Field::new("ab", DataType::Int64, true),
-            Field::new("abc", DataType::Int64, true),
-        ]);
-
-        let input_data = Arc::new(StructArray::new(
-            input_fields.clone(),
-            vec![
-                Arc::new(Int64Array::from(vec![0, 1, 2, 3])),
-                Arc::new(Int64Array::from(vec![4, 7, 10, 11])),
-                Arc::new(Int64Array::from(vec![Some(21), None, Some(387), Some(87)])),
-            ],
-            None,
-        ));
-        let time = Arc::new(TimestampNanosecondArray::from(vec![0, 1, 2, 3]));
-        let subsort = Arc::new(UInt64Array::from(vec![0, 1, 2, 3]));
-        let key_hash = Arc::new(UInt64Array::from(vec![0, 1, 2, 3]));
-        let input_batch = Batch::new_with_data(
-            input_data,
-            time,
-            subsort,
-            key_hash,
-            RowTime::from_timestamp_ns(3),
-        );
-        input_tx.send(input_batch).await.unwrap();
-        std::mem::drop(input_tx);
-
-        execute(
-            "hello".to_owned(),
-            DataType::Struct(input_fields),
-            input_rx,
-            DataType::Struct(output_fields),
-            output_tx,
-        )
-        .await
-        .unwrap();
+//! Execution of physical plans.
+
+use hashbrown::HashMap;
+use sparrow_physical::StepId;
+use sparrow_transforms::TransformPipeline;
+use std::sync::Arc;
+
+use error_stack::ResultExt;
+use itertools::Itertools;
+use sparrow_arrow::Batch;
+use sparrow_core::debug_println;
+use sparrow_scheduler::{PipelineInputs, WorkerPool};
+use uuid::Uuid;
+
+mod error;
+mod source_tasks;
+mod write_channel_pipeline;
+pub use error::*;
+
+use crate::source_tasks::SourceTasks;
+use crate::write_channel_pipeline::WriteChannelPipeline;
+
+pub struct PlanExecutor {
+    worker_pool: WorkerPool,
+    source_tasks: SourceTasks,
+}
 
-        let output = output_rx.recv().await.unwrap();
-        let output = output.data().unwrap().as_struct();
-        let ab = output.column_by_name("ab").unwrap();
-        let abc = output.column_by_name("abc").unwrap();
-        assert_eq!(ab.as_primitive(), &Int64Array::from(vec![4, 8, 12, 14]));
-        assert_eq!(
-            abc.as_primitive(),
-            &Int64Array::from(vec![Some(25), None, Some(399), Some(101)])
-        );
-    }
+const DEBUG_EXECUTOR_CREATION: bool = true;
 
-    /// Execute a physical plan.
-    pub async fn execute(
+impl PlanExecutor {
```

Comment:
This is one of the major changes. This struct (don't love the name) let's us create an executor (using the `sparrow-scheduler::WorkerPool`) from a physical plan. See the tests for it's use.

Reply:
It also feels weird (maybe?) that the usage is:

```
PlanExecutor::try_new(query_id, plan,...).execute()
```

Reply:
I guess it's between this and worker pool, the combination.

1. `WorkerPool` is what lets us add pipelines and execute them. But even that is somewhat split between the "builder" phase and the "execution" phase.
2. `PlanExecutor` just takes the plan, creates the appropriate pipelines, and adds them to the worker pool.

Thinking about the naming together, I wonder if we'd be better with:

1. `WorkerPool` is renamed to `ExecutorBuilder` and `Executor` for the "adding pipelines to a thing" and then "the thing that waits for the pipelines to finish".
2. Then maybe this stays as `PlanExecutor` or `PlanExecutorBuilder` in that it is a layer on top of `ExecutorBuilder` for dealing with plans?